<a href="https://colab.research.google.com/github/ndubuisijosephalx/Data-Science-projects/blob/main/Ndubuisi_J_Assignment_31.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Assignment # 3

In the word_vectors_1 notebook it is shown how to use spacy embedding vectors to find similarity between the words or tokens
Now, use the embedding vectors to find similarity between 2 full sentences

### Q1

Q1
Find the similarity score between the following sentences:

text1 = "i like salty fries and hamburgers"
text2 = "fast food tastes very good"

Note: you need to find the similarity score between the 2 full sentences, and NOT the similarity scores between
the individual words as was shown in the word_vectors_1 notebook. Provide the python code.

### Q2

Q2
Is the similarity score giving a reasonably good indication of the similarity between the 2 sentences?

If your answer is YES, you don't need to give any explanation.

If your answer is NO, give a probable reason why spacy is getting confused. Why is Spacy getting it wrong?

### Q3

Q3
Find the similarity score between the following sentences:

text1 = "river bank"
text2 = "bank account"

Note: you need to find the similarity score between the 2 full sentences, and NOT the similarity scores between
the individual words as was shown in the word_vectors_1 notebook. Provide the python code

### Q4

Q4
Is the similarity score giving a reasonably good indication of the similarity between the 2 sentences?

If your answer is YES, you don't need to give any explanation.

If your answer is NO, give a probable reason why spacy is getting confused. Why is Spacy getting it wrong?

# No. 1 Solution

In [ ]:
import spacy

In [ ]:
#!python -m spacy download en_core_web_lg

In [ ]:

nlp = spacy.load('en_core_web_md')
nlp2 = spacy.load('en_core_web_lg')

Let check if the word 'hamburgers' has a vector and if it exist in our vocabulary. Also we also check for the eucludian for both text1 and text2

## Setences in text1 and text2

In [ ]:
text1 = nlp("i like salty fries and hamburgers")
text2 = nlp("fast food tastes very good")

In [ ]:
for token1,token2 in zip(text1,text2):
    print(token1.text,token2.text)

i fast
like food
salty tastes
fries very
and good


# Check if any of the tokens in text1 and text2 exist in the volcabulary. We will also check for the euclidian norm, and if the tokens has a vector

# Text1  check

In [ ]:
for token in text1:
    print(token.text,token.has_vector,token.vector_norm,token.is_oov)

i True 117.51532 False
like True 50.609623 False
salty True 36.629864 False
fries True 49.507717 False
and True 60.75837 False
hamburgers True 34.503483 False


## Text2 Check

In [ ]:
for tokens in text2:
    print(tokens.text,tokens.has_vector,tokens.vector_norm,tokens.is_oov)

fast True 58.29239 False
food True 75.153 False
tastes True 41.575363 False
very True 61.511013 False
good True 54.353004 False


As we can see the features we just checked in text1 and text2 exist in the pre_trained word embedding model(en_core_web_md). So the medium size word embedding english pre_trained model is enough to address our specific task.

Now that we have known this, let proceed with checking for the two sentences similarity. To explicitly explain the similarity between the two sentences, let us have a check on how each token or word  is related to each other in the sentence.

In [ ]:
for token1 in text1:
    for token2 in text2:
        print(token1.text,token2.text,token1.similarity(token2))

i fast 0.1280231922864914
i food 0.022347047924995422
i tastes 0.14268723130226135
i very 0.12663251161575317
i good 0.20122960209846497
like fast 0.24924501776695251
like food 0.25689175724983215
like tastes 0.3718910813331604
like very 0.36990147829055786
like good 0.5285202860832214
salty fast 0.23733454942703247
salty food 0.40850162506103516
salty tastes 0.5534480810165405
salty very 0.22962722182273865
salty good 0.2958960235118866
fries fast 0.18381565809249878
fries food 0.4907633066177368
fries tastes 0.45439693331718445
fries very 0.3020797669887543
fries good 0.1925533562898636
and fast 0.15189145505428314
and food 0.326991468667984
and tastes 0.05855073034763336
and very 0.3279082179069519
and good 0.1629762351512909
hamburgers fast 0.2480451762676239
hamburgers food 0.46927979588508606
hamburgers tastes 0.47389858961105347
hamburgers very 0.18954713642597198
hamburgers good 0.20561638474464417


As we can see it happens that the word like and good have 52% similarity ans so on

## Now Let Compare the two sentences for similarity

### Medium Pre_trained Model

In [ ]:
import spacy

# Load the spaCy English model
nlp = spacy.load("en_core_web_md")

# Define the sentences
text1 = "i like salty fries and hamburgers"
text2 = "fast food tastes very good"

# Process the sentences with spaCy
doc1 = nlp(text1)
doc2 = nlp(text2)

# Calculate the similarity score
similarity_score = doc1.similarity(doc2)

# Print the similarity score
print("Similarity Score:", similarity_score)

Similarity Score: 0.6227705519265844


### Large Pre_trained Model

In [ ]:
# Load the spaCy English model
nlp2 = spacy.load("en_core_web_lg")

# Define the sentences
text1 = "i like salty fries and hamburgers"
text2 = "fast food tastes very good"

# Process the sentences with spaCy
doc_1 = nlp2(text1)
doc_2 = nlp2(text2)

# Calculate the similarity score
similarity_score = doc_1.similarity(doc_2)

# Print the similarity score
print("Similarity Score:", similarity_score)

Similarity Score: 0.6128936316913924


From the above output, there is approximately 62% similarity between the two sentences for the medium pretrained model and approximately 61% for the large pretrained model, but the question is: is the similarity score reasonable enough to justify their similarity, let us find out in solution 2 below

# Solution 2

My answer is No, 62% or 61% claim is good not enough.

One probable reason why spaCy might be getting confused and providing a misleading similarity score is that the pretrained word vectors used by spaCy, such as the en_core_web_md model, may not adequately capture the semantics and nuances of the given sentences. The pretrained vectors are trained on a large corpus of text, but they might not have specific information about the context or domain of the sentences being compared. Although in our case all the features we are looking for is available in en_core_web_md pretrained model, such as the word embedding vectors, text1 and text2 sentences vocabulary, euclidian norm and so on. For these reasons one may say there is no reason for spacy to be confuse. Let see how true this assertion is.

In this case, the sentences "i like salty fries and hamburgers" and "fast food tastes very good" have some overlapping terms related to food, but they convey different meanings. The first sentence expresses a personal preference for specific types of food, while the second sentence makes a general statement about the taste of fast food. SpaCy's similarity score is based on word vectors and may not be able to accurately capture the overall meaning and intent of the sentences.
## How spacy does this:
spacy focuses on the patterns of word co-occurrences and the context in which words appear together,then assign a vector represntation to each word in the corpus or sentence. The closer words, are assigned a vector similar to each other. 1 shows a very high similiarity and 0 represent no similarity between the two words.

In this case,
The similarity method calculates the similarity score based on the vector representations of the words in the sentences. It compares the vectors for each word in the sentences and aggregates the results to produce an overall similarity score.
The resulting similarity_score is a float value between 0 and 1, where 1 indicates high similarity and 0 indicates no similarity.

To get more accurate results, it is often beneficial to train custom models on domain-specific data or use specialized techniques tailored to the specific task or domain of the sentences being compared.

# No. 3 Solution - Medium Pretrained Model

In [ ]:
import spacy

# Load the spaCy English model
nlp = spacy.load("en_core_web_md")

# Define the sentences
text_1 = "river bank"
text_2 = "bank account"

# Process the sentences with spaCy
doc1 = nlp(text_1)
doc2 = nlp(text_2)

# Calculate the similarity score
similarity_score = doc1.similarity(doc2)

# Print the similarity score
print("Similarity Score:", similarity_score)

Similarity Score: 0.7595661548019256


### Large Pre_trained Model

In [ ]:
nlp3 = spacy.load("en_core_web_lg")
token_3 = nlp3(text_1)
token_4 = nlp3(text_2)
print('Similarity score:',token_3.similarity(token_4))

Similarity score: 0.7595661548019256


# No. 4 Solution

In this context, the similarity score of approximately 76% is moderately reasonable considering the two sentences.

In this case, a score of 0.76 suggests that the two sentences share some similarities in terms of their word usage, context, or meaning. Although the sentences are not identical, they may contain overlapping concepts or associations that contribute to the relatively high similarity score. The discrepancy arises because spaCy's similarity score is based on the word vectors of individual words and their similarity, rather than considering the overall meaning or context of the entire sentence.

In this case, "river bank" and "bank account" have the word "bank" in common, but they have different meanings. "River bank" refers to the edge of a river, while "bank account" refers to a financial account held by a bank. The word "bank" has different contextual associations in each sentence, which spaCy's word vectors may not fully capture